### PR TITLE
interface: update crp to support override

### PR DESCRIPTION
**Pull request by zhiying-lin**
### Description of your changes

Update the CRP to support override feature.

Fixes #

I have:

- [X] Run `make reviewable` to ensure this PR is ready for review.

### Special notes for your reviewer

The override snapshot and other internal CRD will be created in a separate PR.

### CODE REVIEW

File: apis/placement/v1beta1/commons.go [hunk 1]
```diff
@@ -81,3 +81,29 @@ const (
...
+	// +required
+	Name string `json:"name"`
+	// Namespace of the override snapshot.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+	// Type of override snapshot. Can be "ClusterResourceOverride", or "ResourceOverride". Default is ClusterResourceOverride.
+	// +kubebuilder:validation:Enum=ClusterResourceOverride;ResourceOverride
+	// +kubebuilder:default=ClusterResourceOverride
+	// +optional
+	Type OverrideType `json:"type,omitempty"`
```
Comment (ryanzhang-oss) [inline apis/placement/v1beta1/commons.go:97]: given that one has namespace value while the other doesn't, do we still need this field?

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 2]
```diff
@@ -513,6 +568,15 @@ const (
...
 
+	// ResourceOverriddenConditionType indicates whether all the selected resources have been overridden successfully
+	// before applying to the target cluster if there is any override defined.
+	// Its condition status can be one of the following:
+	// - "True" means all the selected resources are successfully overridden before applying to the target cluster or
+	// override is not needed if there is no override defined with the reason of NoOverrideSpecified.
+	// - "False" means some of them have failed.
+	// - "Unknown" means we haven't finished the override yet.
+	ResourceOverriddenConditionType ResourcePlacementConditionType = "ResourceOverridden"
+
```
Comment (ryanzhang-oss) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:579]: where is this condition supposed to apply to? We don't have a resourcePlacement yet.

Comment (zhiying-lin) [reply apis/placement/v1beta1/clusterresourceplacement_types.go:579]: It's an new condition of the existing ResourcePlacementStatus https://github.com/Azure/fleet/blob/main/apis/placement/v1beta1/clusterresourceplacement_types.go#L444

File: apis/placement/v1beta1/commons.go [hunk 3]
```diff
@@ -81,3 +81,29 @@ const (
 	// This is used to remember if an "unscheduled" binding was moved from a "bound" state or a "scheduled" state.
 	PreviousBindingStateAnnotation = fleetPrefix + "previous-binding-state"
 )
+
+// OverrideSnapshotIdentifier defines the identity of an override snapshot.
+type OverrideSnapshotIdentifier struct {
```
Comment (zhiying-lin) [inline apis/placement/v1beta1/commons.go:86]: will move this struct to v1alpah1 api once #661 is merged

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 4]
```diff
@@ -442,6 +484,10 @@ type ResourcePlacementStatus struct {
 	// Conditions is an array of current observed conditions for ResourcePlacementStatus.
 	// +optional
 	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// Overrides contains a list of override snapshot associated with the selected resources.
+	// +optional
+	Overrides []OverrideSnapshotIdentifier `json:"overrides,omitempty"`
```
Comment (circy9) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:490]: Shall we just separate them to make them more explicit?

ApplicableResourceOverrides []NamespacedName

ApplicableClusterResourceOverrides []string

type NamespacedName struct {
  namespace
  name
}

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 5]
```diff
@@ -477,6 +523,15 @@ const (
...
 	ClusterResourcePlacementSynchronizedConditionType ClusterResourcePlacementConditionType = "ClusterResourcePlacementSynchronized"
 
+	// ClusterResourcePlacementOverriddenConditionType indicates whether all the selected resources have been overridden
+	// successfully before applying to the target cluster.
+	// Its condition status can be one of the following:
+	// - "True" means all the selected resources are successfully overridden before applying to the target cluster or
+	// override is not needed if there is no override defined with the reason of NoOverrideSpecified.
+	// - "False" means some of them have failed. We will place some detailed failure in the FailedResourcePlacement array.
+	// - "Unknown" means we haven't finished the override yet.
+	ClusterResourcePlacementOverriddenConditionType ClusterResourcePlacementConditionType = "ClusterResourcePlacementOverridden"
```
Comment (circy9) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:533]: Why is this one ClusterResourcePlacementOverridden but the other one is ResourceOverridden? Do we really need "placement" in between?

Just ClusterResourceOverridden?

File: apis/placement/v1beta1/commons.go (hunk 1 above)
Comment (circy9) [reply apis/placement/v1beta1/commons.go:97]: Shall we just separate them into two arrays? We don't need a single data structure then. See my comments above.

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 6]
```diff
@@ -513,6 +568,15 @@ const (
...
 	ResourceWorkSynchronizedConditionType ResourcePlacementConditionType = "WorkSynchronized"
 
+	// ResourceOverriddenConditionType indicates whether all the selected resources have been overridden successfully
+	// before applying to the target cluster if there is any override defined.
+	// Its condition status can be one of the following:
+	// - "True" means all the selected resources are successfully overridden before applying to the target cluster or
+	// override is not needed if there is no override defined with the reason of NoOverrideSpecified.
+	// - "False" means some of them have failed.
+	// - "Unknown" means we haven't finished the override yet.
+	ResourceOverriddenConditionType ResourcePlacementConditionType = "ResourceOverridden"
```
Comment (circy9) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:578]: What is the difference between this one and ClusterResourcePlacementOverridden? Will they both used in one placement?

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 7]
```diff
@@ -282,6 +282,48 @@ type RolloutStrategy struct {
...
+	// is owned by non-CRP appliers.
+	ApplyStrategyTypeServerSideApply ApplyStrategyType = "ServerSideApply"
+)
+
+// ServerSideApplyConfig defines the configuration for server side apply.
+// https://kubernetes.io/docs/reference/using-api/server-side-apply/
+type ServerSideApplyConfig struct {
+	// Force represents to force apply to succeed when resolving the conflicts.
+	// +optional
+	ForceConflicts bool `json:"force"`
```
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply apis/placement/v1beta1/clusterresourceplacement_types.go:326]: not accurate for the non-conflicting fields. Updated.

File: apis/placement/v1beta1/commons.go [hunk 8]
```diff
@@ -81,3 +81,24 @@ const (
...
+	// +required
+	Name string `json:"name"`
+	// Namespace is namespace of the namespaced scope resource.
+	// +required
+	Namespace string `json:"namespace"`
+}
+
+// OverrideType identifies the type of override.
+// +enum
+type OverrideType string
```
Comment (circy9) [inline apis/placement/v1beta1/commons.go:97]: Please delete as this is no longer needed.

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 9]
```diff
@@ -431,6 +482,16 @@ type ResourcePlacementStatus struct {
 	// +optional
 	ClusterName string `json:"clusterName,omitempty"`
 
+	// ApplicableResourceOverrides contains a list of applicable ClusterResourceOverride snapshots associated with the
```
Comment (circy9) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:485]: s/ClusterResourceOverride/ResourceOverride/

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 10]
```diff
@@ -431,6 +482,16 @@ type ResourcePlacementStatus struct {
...
+	// selected resources.
+	// +optional
+	ApplicableResourceOverrides []NamespacedName `json:"applicableResourceOverrides,omitempty"`
+
+	// ApplicableClusterResourceOverrides contains a list of applicable ResourceOverride snapshots associated with the
```
Comment (circy9) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:490]: s/ResourceOverride/ClusterResourceOverride/

**Review by circy9 — CHANGES_REQUESTED** [review]

File: apis/placement/v1beta1/commons.go (hunk 8 above)
Comment (zhiying-lin) [reply apis/placement/v1beta1/commons.go:97]: good catch! thanks

**Review by circy9 — APPROVED** [review]

**Pull request merged**